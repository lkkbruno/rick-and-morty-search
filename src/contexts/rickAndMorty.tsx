import React, { createContext, useState, useEffect, PropsWithChildren } from 'react';
import axios from 'axios';

export interface ICharacter {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  origin: {
    name: string;
    url: string;
  };
  location: {
    name: string;
    url: string;
  };
  image: string;
  episode: string[];
  url: string;
  created: string;
}

interface RickAndMortyContextData {
  characters: ICharacter[];
  searchTerm: string;
  setSearchTerm: (term: string) => void;
}

export const RickAndMortyContext = createContext<RickAndMortyContextData>({
  characters: [],
  searchTerm: '',
  setSearchTerm: () => { },
});

export const RickAndMortyProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [characters, setCharacters] = useState<ICharacter[]>([]);
  const [searchTerm, setSearchTerm] = useState<string>('');

  const fetchCharacters = () => {
    axios
      .get(`https://rickandmortyapi.com/api/character`, {
        params: {
          name: searchTerm,
        }
      })
      .then((response) => {
        setCharacters(response.data.results);
      }).catch(() => {
        setCharacters([]);
      });
  }

  useEffect(() => {
    fetchCharacters();
  }, [searchTerm]);

  return (
    <RickAndMortyContext.Provider value={{ characters, searchTerm, setSearchTerm }}>
      {children}
    </RickAndMortyContext.Provider>
  );
};
