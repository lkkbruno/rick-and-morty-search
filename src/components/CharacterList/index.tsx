import React, { useContext, useState, useEffect } from 'react';
import styled from 'styled-components';
import { RickAndMortyContext, ICharacter } from '../../contexts/rickAndMorty';
import Character from '../Character';

const List = styled.ul`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  list-style: none;
  padding: 0;
  margin: 0;
`;

interface CharacterListProps {
  characters: ICharacter[],
}

const CharacterList: React.FC<CharacterListProps> = ({ characters }) => {
  const { searchTerm } = useContext(RickAndMortyContext);
  const [filteredCharacters, setFilteredCharacters] = useState<ICharacter[]>(characters);

  useEffect(() => {
    setFilteredCharacters(
      characters.filter((character) =>
        character.name.toLowerCase().includes(searchTerm.toLowerCase())
      )
    );
  }, [characters, searchTerm]);

  return (
    <List>
      {filteredCharacters.map((character) => (
        <Character key={character.id} character={character} />
      ))}
    </List>
  );
};

export default CharacterList;
