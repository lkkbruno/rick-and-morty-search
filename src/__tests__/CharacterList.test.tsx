import React from 'react';
import { render, waitFor, screen } from '@testing-library/react';
import CharacterList from '../components/CharacterList';
import { ICharacter, RickAndMortyProvider, RickAndMortyContext } from '../contexts/rickAndMorty';
import { rest } from 'msw';
import { server } from './mocks/setupServer';
import { mockCharactersList } from './mocks/characters'

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('CharacterList', () => {
  it('should match snapshot', async () => {
    const { container } = render(<CharacterList characters={mockCharactersList} />);
    expect(container).toMatchSnapshot();
  });

  it('should render a list of characters', async () => {
    render(
      <CharacterList characters={mockCharactersList} />
    );

    await waitFor(() => expect(screen.getByText('Rick Sanchez')).toBeInTheDocument());
    expect(screen.getByText('Morty Smith')).toBeInTheDocument();
  });
});
