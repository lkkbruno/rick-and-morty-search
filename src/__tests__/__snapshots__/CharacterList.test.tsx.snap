// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterList should match snapshot 1`] = `
<div>
  <ul
    class="sc-gsFSXq eQVcDd"
  >
    <li
      class="sc-gEvEer cpkkRY"
    >
      <img
        alt="Rick Sanchez"
        class="sc-eqUAAy cQJqLn"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      />
      <div
        class="sc-fqkvVR bFKHFo"
      >
        <h2
          class="sc-dcJsrY kWDMGg"
        >
          Rick Sanchez
        </h2>
        <p
          class="sc-iGgWBj hNqgHF"
        >
          Gender: Male
        </p>
        <p
          class="sc-iGgWBj hNqgHF"
        >
          Status: Alive
        </p>
        <p
          class="sc-iGgWBj hNqgHF"
        >
          Species: Human
        </p>
      </div>
    </li>
    <li
      class="sc-gEvEer cpkkRY"
    >
      <img
        alt="Morty Smith"
        class="sc-eqUAAy cQJqLn"
        src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
      />
      <div
        class="sc-fqkvVR bFKHFo"
      >
        <h2
          class="sc-dcJsrY kWDMGg"
        >
          Morty Smith
        </h2>
        <p
          class="sc-iGgWBj hNqgHF"
        >
          Gender: Male
        </p>
        <p
          class="sc-iGgWBj hNqgHF"
        >
          Status: Alive
        </p>
        <p
          class="sc-iGgWBj hNqgHF"
        >
          Species: Human
        </p>
      </div>
    </li>
  </ul>
</div>
`;
